theano/scalar/basic.py

```diff
@@ -3807,3 +3812,79 @@ def __setstate__(self, d):
         self.init_fgraph()
         self.init_py_impls()
         assert self._c_code
+
+
+class Compositef32(object):
+    # This is a dict of scalar op classes that need special handling
+    special = {}
+
+    def apply(self, fgraph):
+        mapping = {}
+        topo = fgraph.toposort()
+        for i in fgraph.inputs:
+            if i.dtype == 'float16':
+                mapping[i] = get_scalar_type('float32')()
+            else:
+                mapping[i] = i
+        for node in topo:
+            # Patch up for constants
+            for i in node.inputs:
+                if i not in mapping:
+                    assert type(i) is ScalarConstant
+                    mapping[i] = i
```

Comment:
What if the ScalarConstant is float16? I think in that case, it should be converted to float32.